DigitalLearningSolutions.Web/Attributes/SetApplicationTypeAttribute.cs

```diff
@@ -0,0 +1,56 @@
+﻿namespace DigitalLearningSolutions.Web.Attributes
+{
+    using System;
+    using DigitalLearningSolutions.Web.Models.Enums;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.AspNetCore.Mvc.Filters;
+
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method)]
+    public class SetApplicationTypeAttribute : Attribute, IActionFilter
+    {
+        private readonly string? applicationRouteParamName;
+        private readonly string? applicationTypeName;
+
+        public SetApplicationTypeAttribute(string? applicationTypeName = null, string? applicationRouteParamName = null)
+        {
+            this.applicationTypeName = applicationTypeName;
+            this.applicationRouteParamName = applicationRouteParamName;
+            if (applicationTypeName != null && applicationRouteParamName != null)
+            {
+                throw new Exception(
+                    "Only one argument may be passed to SetApplicationTypeAttribute. " +
+                    "Define applicationTypeName if the controller is only used in one application, " +
+                    "or define applicationRouteParamName if the controller is used across multiple applications."
+                );
+            }
+
+            if (applicationTypeName == null && applicationRouteParamName == null)
+            {
+                throw new Exception(
+                    "One argument must be passed to SetApplicationTypeAttribute. " +
+                    "Define applicationTypeName if the controller is only used in one application, " +
+                    "or define applicationRouteParamName if the controller is used across multiple applications."
+                );
+            }
+        }
+
+        public void OnActionExecuted(ActionExecutedContext context) { }
+
+        public void OnActionExecuting(ActionExecutingContext context)
+        {
+            if (!(context.Controller is Controller controller))
+            {
+                return;
+            }
+
+            if (applicationTypeName != null)
+            {
+                controller.ViewData["ApplicationType"] = (ApplicationType)applicationTypeName;

```

Comment:
Now I've had a bit more time to consider it, I'm wary about `applicationType` as a name - it's not really talking about types of application like game, browser, word processor; it's a selection of applications. How do you feel about it? Does `subApplication` seem like a reasonable alternative?

Reply:
Yeah that all makes sense. I've changed ApplicationType to DlsSubApplication everywhere it is used, happy with that :)